lib/io/pwm.cpp

```diff
@@ -0,0 +1,214 @@
+#include "pwm.hpp"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+namespace hyped::io {
+
+std::optional<Pwm> Pwm::create(core::Logger &logger, const PwmModule pwm_module)
+{
+  Pwm pwm(logger, pwm_module);
+  const auto initialisation_result = pwm.initialise();
+  if (initialisation_result == core::Result::kFailure) {
+    pwm.logger_.log(core::LogLevel::kFatal, "Failed to initialise PWM");
+    return std::nullopt;
+  }
+  pwm.logger_.log(core::LogLevel::kDebug, "Successfully initialised PWM");
+  return pwm;
+}
```

Comment:
by the time the object is creates we should have four non-zero file descriptors; this would allow you to make the descriptors `const`

so, here you should open each file individually and then call a constructor that takes the logger, the pwm_module, and the descriptors